src/librbd/mirror_snapshot/PromoteRequest.h

```diff
@@ -0,0 +1,69 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```

Comment:
At the moment when I was implementing `Promote/DemoteRequest` I was not sure how it would end up. And even now, for example, I am not sure about the force promotion. I think we will want to rollback to the latest copied snapshot. Is `PromoteRequest` supposed to do this or do we expect the user does it manually (or with some additional command)?

Anyway, even if they do not differ much I like more this way instead of having `PromoteDemoteRequest(bool promote)`. But I will be fine with any suggestions.

Reply:
True -- a forced promotion would need to rollback to the last fully committed snapshot. We would, however, need to ensure that the local `rbd-mirror` daemon process isn't syncing the data before the rollback is started. That might involve creating the forced promotion snapshot and waiting for the image watchers (for `rbd-mirror`) to disappear before starting the rollback.

Reply:
Do we need some special field to distinguish that 'forced promotion snapshot' from other primary snapshots?

If the process crashes after we created the forced promotion snapshot and before the rollback is complete, how the user (e.g. the restarted promotion) could detect this state? I suppose we need to create yet another promotion snapshot after the rollback is complete. Still we need to somehow distinguish between the "forced" primary snapshot created at the start and the "normal" primary snapshot created at the end? Add yet another boolean (`forced`) to the primary snapshot namespace?

Reply:
If you create a non-primary snapshot first w/ the mirror uuid set to "<orphan>" or just "", you get the same effect in that you have now broken the link between the other site's primary and the forced promotion. Then you could re-use the complete and last object copied fields to record the rollback upon forced promotion. When the rollback is complete, you can add the primary snapshot.